workbench/src/sidebar.c

```diff
@@ -0,0 +1,1164 @@
+			if (current == prj)
+			{
+				return TRUE;
+			}
+		}while (gtk_tree_model_iter_next(model, iter));
+	}
+	return FALSE;
+}
+
+
+/* Remove all rows from the side bar tree with the given data id */
+static void sidebar_remove_nodes_with_data_id(guint toremove, GtkTreeIter *start)
+{
+	GtkTreeIter iter;
+	GtkTreeModel *model;
+	gboolean has_next;
+	guint dataid;
+
+	model = gtk_tree_view_get_model(GTK_TREE_VIEW(sidebar.file_view));
+	if (start == NULL)
+	{
+		if (!gtk_tree_model_get_iter_first(model, &iter))
+		{
+			return;
+		}
+	}
+	else
+	{
+		iter = *start;
+	}
+
+	do
+	{
+		gtk_tree_model_get(model, &iter, FILEVIEW_COLUMN_DATA_ID, &dataid, -1);
+		if (dataid == toremove)
+		{
+			has_next = gtk_tree_store_remove(sidebar.file_store, &iter);
+		}
+		else
+		{
+			has_next = gtk_tree_model_iter_next(model, &iter);
+		}
+	}
+	while (has_next);
+}
+
+
+/* Insert all project bookmarks into the sidebar file tree */
+static void sidebar_insert_project_bookmarks(WB_PROJECT *project, GtkTreeIter *parent, gint *position)
+{
+	GIcon *icon;
+	guint index, max;
+	GtkTreeIter iter;
+
+	if (project == NULL)
+		return;
+
+	max = wb_project_get_bookmarks_count(project);
+	if (max == 0)
+		return;
+
+	icon = g_icon_new_for_string("workbench-bookmark", NULL);
+	for (index = 0 ; index < max ; index++)
+	{
+		gchar *file, *name;
+
+		file = wb_project_get_bookmark_at_index(project, index);
+		name = get_any_relative_path(wb_project_get_filename(project), file);
+		gtk_tree_store_insert_with_values(sidebar.file_store, &iter, parent, *position,
+			FILEVIEW_COLUMN_ICON, icon,
+			FILEVIEW_COLUMN_NAME, name,
+			FILEVIEW_COLUMN_DATA_ID, DATA_ID_PRJ_BOOKMARK,
+			FILEVIEW_COLUMN_ASSIGNED_DATA_POINTER, file,
+			-1);
+		(*position)++;
+	}
+	g_object_unref(icon);
+}
+
+
+/* Update the sidebar for a project, only update title */
+static void sidebar_update_project_title(WB_PROJECT *project)
+{
+	GtkTreeIter iter;
+
+	if (wb_globals.opened_wb == NULL)
+		return;
+
+	if (sidebar_get_project_iter(project, &iter))
+	{
+		gint length;
+		gchar text[200];
+
+		length = g_snprintf(text, sizeof(text), "%s", wb_project_get_name(project));
+		if (length < (sizeof(text)-1) && wb_project_is_modified(project))
+		{
+			text [length] = '*';
+			text [length+1] = '\0';
+		}
+		gtk_tree_store_set(sidebar.file_store, &iter,
+			FILEVIEW_COLUMN_NAME, text,
+			-1);
+	}
+}
+
+
+/* Update the sidebar for a project only */
+static void sidebar_update_project(WB_PROJECT *project)
+{
+	GtkTreeIter iter;
+
+	if (wb_globals.opened_wb == NULL)
+		return;
+
+	if (sidebar_get_project_iter(project, &iter))
+	{
+		gint length, position;
+		gchar text[200];
+
+		length = g_snprintf(text, sizeof(text), "%s", wb_project_get_name(project));
+		if (length < (sizeof(text)-1) && wb_project_is_modified(project))
+		{
+			text [length] = '*';
+			text [length+1] = '\0';
+		}
+		gtk_tree_store_set(sidebar.file_store, &iter,
+			FILEVIEW_COLUMN_NAME, text,
+			-1);
```

Comment:
True. Merged the functions together and added a parameter to choose to only update the project title or all parts of the project.